### DISCUSSION TITLE
GitHub Package Registry - NuGet - Search API not working correctly

### DISCUSSION THREAD

**Discussion by Falco20019** [community/community#101961] [Packages]
### Select Topic Area

Bug

### Body

According to the [specification](https://learn.microsoft.com/en-us/nuget/api/search-query-service-resource), the total hits should always include the full amount, independant of take and skip (to allow paging).

[On NuGet.org](https://azuresearch-usnc.nuget.org/query?q=grpc&skip=0&take=1&prerelease=true&semVerLevel=2.0.0) this works fine. `TotalHits` is always `1145` and changing `skip` from `0` to `1` will skip the first entry.

[On GitHub Packages](https://nuget.pkg.github.com/Falco20019/query?q=grpc&skip=0&take=1&prerelease=true&semVerLevel=2.0.0) this does not work at all. `TotalHits` is always less or equal to `take` and changing `skip` from `0` to `1` will do nothing at all. Paging is not possible due to that.

**Comment by Falco20019** [comment]
According to the GitHub Enterprise support, this is a known bug. There the workaround suggested by them:

* Use the GitHub REST API endpoint `/orgs/{org}/packages` to retrieve a list of nuget packages for the organization and leverage its pagination capability (using the `page` and `per_page` query string parameters)
* Use the GitHub REST API endpoint `/orgs/{org}/packages/{package_type}/{package_name}/versions` to retrieve a list of versions for each package and leverage its pagination capability (using the `page` and `per_page` query string parameters)
* Use the GitHub Packages Nuget Registry API endpoint `/{org}/{package_name}/{package_version}.json`  to retrieve metadata for a package version
  * This allows for retrieval of the `.catalogEntry.description` attribute which was your requirement that couldn't be met using either the REST API or the GraphQL API.

**Comment by Lanrad23** [comment]
I've run into the same problem too. It's really annoying when the API doesn't do what it's supposed to, like letting us go through a list page by page with the totalHits, take, and skip settings. This bug makes it hard to look through all the NuGet packages on GitHub Packages because the totalHits doesn't show the real number of packages, and changing the skip setting doesn't actually skip anything like it should.

**Comment by Falco20019** [comment]
Yep, especially if you use tooling that relies on it to work and does not implement a specific vendor workaround...

**Comment by Latency** [comment]
This works!  Stick it in the directory just above the root of your self-hosted runner.

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="github" value="https://nuget.pkg.github.com/<USER>/index.json" protocolVersion="3" />
  </packageSources>
  <activePackageSource>
    <add key="All" value="(Aggregate source)" />
  </activePackageSource>
  <packageRestore>
    <add key="enabled" value="True" />
    <add key="automatic" value="True" />
  </packageRestore>
  <bindingRedirects>
    <add key="skip" value="False" />
  </bindingRedirects>
  <packageManagement>
    <add key="format" value="1" />
    <add key="disabled" value="False" />
  </packageManagement>
  <config>
    <add key="defaultPushSource" value="https://nuget.pkg.github.com/<USER>/index.json" />
  </config>
  <packageSourceCredentials>
    <github>
        <add key="Username" value="<USER>" />
        <add key="ClearTextPassword" value="<TOKEN>" />
      </github>
  </packageSourceCredentials>
</configuration>
```

Output:
```
PS D:\Runners\actions-runner-AsyncTask\work\AsyncTask\AsyncTask> nuget search "AsyncTask" -Source "github"
====================
Source: github
--------------------
> AsyncTask | 3.2.1 | Downloads: 0
  Asynchronous Task Extension Library
--------------------
```

**Comment by lxop** [comment]
In addition to `skip` and `take` not behaving properly, `q` doesn't behave properly either. The spec says
>If no q is provided, all packages should be returned, within the boundaries imposed by skip and take

but I am seeing that when `q` is omitted (or set to an empty string), the server only responds with packages created "recently" (I haven't determined exactly how recently, but we have a package created a little over 6 months ago that isn't being returned so that puts an upper bound on it). We have 17 packages, but only 2 are being returned.

Interestingly, if I omit _all_ params from the query (no `q`, no `take`, no `skip`, nothing), the server does return all the packages as expected. Unfortunately this doesn't help since I want Visual Studio to list available packages, and I can't change how VS constructs its query params.
